src/lib/module.ts

```diff
@@ -20,21 +20,25 @@ import {filter} from 'rxjs/operators/filter';
 import {take} from 'rxjs/operators/take';
 
 import {EventReplayer} from './api/event.replayer';
-import {getInlinePrebootCode} from './api/inline.preboot.code';
+import {assign, defaultOptions, getInlinePrebootCode} from './api/inline.preboot.code';
 import {PrebootOptions} from './common/preboot.interfaces';
 import {PREBOOT_NONCE} from './common/tokens';
 
 
 export const PREBOOT_OPTIONS = new InjectionToken<PrebootOptions>('PrebootOptions');
 
 export function prebootHook(doc: Document,
-                            prebootOpts: PrebootOptions,
+                            customOpts: PrebootOptions,
                             nonce: string|null,
                             platformId: Object,
                             appRef: ApplicationRef,
                             eventReplayer: EventReplayer) {
   // necessary because of angular/angular/issues/14485
   const res = () => {
+    // Use `Object.create(null)` as a base as otherwise user can modify
+    // the prototype of the options object.
+    const prebootOpts = <PrebootOptions>assign(Object.create(null), defaultOptions, customOpts);
```

Comment:
This was a good idea, but in the future, since this is in the Angular module part of the app (and thus gets transpiled by the end user), you can just do an object spread here.

`const prebootOpts = {...defaultOptions, ...customOpts};`

Reply:
Good idea, I’ll update it.

Reply:
Actually, I can't do it as TypeScript's implementation of object spread is nonconformant and does regular assignments instead of `Object.defineProperty` and suffers from the same `__proto__` prototype mutation issue as `Object.assign` with a regular object. :(

See http://www.typescriptlang.org/play/#src=var%20o%20%3D%20Object.create(null)%3B%0D%0A%0D%0Ao.a%20%3D%20'foo'%3B%0D%0Ao.__proto__%20%3D%20%5B%5D%3B%0D%0A%0D%0Avar%20o1%20%3D%20Object.assign(%7B%7D%2C%20o)%3B%0D%0A%0D%0Avar%20o2%20%3D%20%7B...o%7D%3B%0D%0A%0D%0Aconsole.log('o1%20proto%20array%3F%20(expected%3A%20true)'%2C%20Array.isArray(Object.getPrototypeOf(o1)))%3B%0D%0Aconsole.log('o2%20proto%20array%3F%20(expected%3A%20false)'%2C%20Array.isArray(Object.getPrototypeOf(o2)))%3B%0D%0A

Reply:
I think if someone goes to the trouble to typecast an object to `PrebootOptions` and replace the `__proto__`, maybe they deserve to cause errors?